fpga_interchange/pseudo_pip_model.cc

```diff
@@ -0,0 +1,475 @@
+            continue;
+        }
+
+        PipId pip;
+        pip.tile = tile;
+        pip.index = pip_idx;
+        const std::vector<size_t> &sites = ctx->pseudo_pip_data.get_possible_sites_for_pip(ctx, pip);
+
+        int32_t site_for_pip = -1;
+        for(size_t possible_site : sites) {
+            if(active_sites.count(possible_site)) {
+                site_for_pip = possible_site;
+                break;
+            }
+        }
+
+        if(site_for_pip < 0) {
+            site_for_pip = sites.at(0);
+        }
+
+        pseudo_pip_sites[pip_idx] = site_for_pip;
+        site_to_pseudo_pips[site_for_pip].push_back(pip_idx);
+    }
+
+    for(auto & site_pair : site_to_pseudo_pips) {
+        update_site(ctx, site_pair.first);
+    }
+}
+
+bool PseudoPipModel::checkPipAvail(const Context *ctx, PipId pip) const {
+    bool allowed = allowed_pseudo_pips.get(pip.index);
+    if(!allowed) {
+#ifdef DEBUG_PSEUDO_PIP
+        if(ctx->verbose) {
+            log_info("Pseudo pip %s not allowed\n", ctx->nameOfPip(pip));
+        }
+#endif
+    }
+
+    return allowed;
+}
+
+void PseudoPipModel::bindPip(const Context *ctx, PipId pip) {
+    // If pseudo_pip_sites is empty, then prepare_for_routing was never
+    // invoked.  This is likely because PseudoPipModel was constructed during
+    // routing.
+    if(pseudo_pip_sites.empty()) {
+        prepare_for_routing(ctx, ctx->tileStatus.at(tile).sites);
+    }
+
+    // Do not allow pseudo pips to be bound if they are not allowed!
+    NPNR_ASSERT(allowed_pseudo_pips.get(pip.index));
+
+    // Mark that this pseudo pip is active.
+    auto result = active_pseudo_pips.emplace(pip.index);
+    NPNR_ASSERT(result.second);
+
+    // Update the site this pseudo pip is within.
+    size_t site = pseudo_pip_sites.at(pip.index);
+    update_site(ctx, site);
+}
+
+void PseudoPipModel::unbindPip(const Context *ctx, PipId pip) {
+    // It should not be possible for unbindPip to be invoked with
+    // pseudo_pip_sites being empty.
+    NPNR_ASSERT(!pseudo_pip_sites.empty());
+
+    NPNR_ASSERT(active_pseudo_pips.erase(pip.index));
+
+    // Remove the site this pseudo pip is within.
+    size_t site = pseudo_pip_sites.at(pip.index);
+    update_site(ctx, site);
+}
+
+void PseudoPipModel::update_site(const Context *ctx, size_t site) {
+    // update_site consists of several steps:
+    //
+    //  - Find all BELs within the site used by pseudo pips.
+    //  - Trivially marking other pseudo pips as unavailable if it requires
+    //    logic BELs used by active pseudo pips (or bound by cells).
+    //  - Determine if remaining pseudo pips can be legally placed.  This
+    //    generally consists of:
+    //     - Checking LUT element
+    //     - FIXME: Checking constraints (when metadata is available)
+
+    const std::vector<int32_t> pseudo_pips_for_site = site_to_pseudo_pips.at(site);
+
+    // unused_pseudo_pips is a scratch array kept in the model, but is intended to be used during
```

Comment:
I think a word is missing from the end of this comment

Reply:
Yep.  Fixed.